src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -497,34 +497,75 @@ int encrypt(byte[] in, int inOfs, int len) {
-     * a multiple of the block size
      * @return the number of bytes placed into the <code>out</code> buffer
      */
-    int encrypt(byte[] in, int inOfs, int len, byte[] out, int outOfs) {
-        checkDataLength(processed, len);
+    int encrypt(byte[] in, int inOfs, int inLen, byte[] out, int outOfs) {
+        checkDataLength(processed, inLen);
 
         processAAD();
-        int remainder = len % blockSize;
-        if ( remainder > 0) {
+        // 'len' stores the length to use with buffer 'in'.
+        // 'inLen' stores the length returned by the method.
+        int len = inLen;
+
+        // if there is enough data in the ibuffer and 'in' construct a block
+        // to encrypt
+
+        if (ibuffer != null && ibuffer.size() > 0 &&
+            (inLen + ibuffer.size() >= blockSize)) {
+            ArrayUtil.nullAndBoundsCheck(in, inOfs, inLen);
+            ArrayUtil.nullAndBoundsCheck(out, outOfs, inLen);
+            byte[] block = new byte[blockSize];
+            int inLenUsed = constructBlock(ibuffer.toByteArray(), in, block);
+            gctrPAndC.update(block, 0, blockSize, out, outOfs);
+            processed += blockSize;
+            ghashAllToS.update(out, outOfs, blockSize);
+            inOfs += inLenUsed;
+            inLen -= inLenUsed;
+            len += (blockSize - inLenUsed);
+            outOfs += blockSize;
+            ibuffer.reset();
+        }
+
+        // if there it not enough data in the ibuffer and 'in', then append
+        // that to the ibuffer.
+        int remainder = inLen % blockSize;
+        if (remainder > 0) {
             if (ibuffer == null) {
-                ibuffer = new ByteArrayOutputStream(len % blockSize);
+                ibuffer = new ByteArrayOutputStream(inLen % blockSize);
             }
             len -= remainder;
-            ibuffer.write(in, len, remainder);
+            inLen -= remainder;
+            // remainder offset is based on original buffer length
+            ibuffer.write(in, inOfs + inLen, remainder);
         }
+
         if (len > 0) {
-            ArrayUtil.nullAndBoundsCheck(in, inOfs, len);
-            ArrayUtil.nullAndBoundsCheck(out, outOfs, len);
+            ArrayUtil.nullAndBoundsCheck(in, inOfs, inLen);
+            ArrayUtil.nullAndBoundsCheck(out, outOfs, inLen);
 
-            gctrPAndC.update(in, inOfs, len, out, outOfs);
-            processed += len;
-            ghashAllToS.update(out, outOfs, len);
+            gctrPAndC.update(in, inOfs, inLen, out, outOfs);
+            processed += inLen;
+            ghashAllToS.update(out, outOfs, inLen);
         }
 
         return len;
     }
 
+    // Returns length of bytes taken from 'in' buffer
+    int constructBlock(byte[] buffer, byte[] in, byte[] block) {
+        int buflen = buffer.length;
+        if (buflen >= blockSize) {
+            System.arraycopy(buffer, 0, block, 0, blockSize);
+            buflen -= block.length;
+            return 0;
```

Comment:
Will bufLen > blockSize? Judging from the context of this method, buffer.length should always <= blockSize and  never be larger? If bufLen == blockSize, perhaps we can just use buffer directly and no need to copy.

Reply:
This method is being removed

Reply:
Ok.